### PR TITLE
Import Apply infrastructure from Approved Premises

### PR DESCRIPTION
**Pull request by bryangaledxw**
# Changes in this PR

This PR imports the Apply infrastructure from AP largely unchanged. The intent is to give us the base from which to start building TA's "Apply" feature, which will operate similarly

The bulk of the changes compared to AP are to support TA's RestClient taking a CallConfig object, rather than a token string. To prevent the dependency tree exploding, getSections is also changed to return an empty array.

Commit are roughly in utils, clients, services, controllers, order, with some variation to keep tests passing a dependencies satisfied at each point. It is noted in each commit where we are deviating from AP.

This PR does not bring over any Apply integration or E2E testing code.

# Release checklist

As part of our continuous deployment strategy we must ensure that this work is
ready to be released at any point. Before merging to `main` we must first
confirm:

## Pre merge checklist

- [ ] Have all changes to any dependencies been deployed to both `preprod` and
    `production` in a backwards compatible way? (This includes our API,
    infrastructure, third-party integrations and any secrets or environment variables)
- [ ] Has a required data migration been included in this change or been done in
    advance?

## Post merge checklist

Once we've merged we now need to release this through to production before
considering it done.

[Find the build-and-deploy job in CircleCI](https://app.circleci.com/pipelines/github/ministryofjustice/hmpps-temporary-accommodation-ui)
and work through the following steps:

- [ ] Has the product manager asked to provide approval for this feature?
  - [ ] Has the product manager approved?
- [ ] Manually approve release to preprod
- [ ] Verify change released to preprod
- [ ] Manually approve release to prod
- [ ] Verify change released to production

Should a release fail at any step, you as the author should now lead the work to
fix it as soon as possible.

### CODE REVIEW

File: server/utils/assessments/isAssessment.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+import {
+  ApprovedPremisesApplication as Application,
+  ApprovedPremisesAssessment as Assessment,
+} from '../../@types/shared'
+
+export default (applicationOrAssessment: Application | Assessment): applicationOrAssessment is Assessment => {
+  return 'allocatedToStaffMemberId' in applicationOrAssessment
```
Comment (pezholio) [inline server/utils/assessments/isAssessment.ts:7]: This has just changed (ref https://github.com/ministryofjustice/hmpps-approved-premises-ui/pull/546) as an assessment no longer has an `allocatedToStaffMemberId`

File: package.json [hunk 2]
```diff
@@ -33,6 +33,7 @@
     "api-stubs:reset": "npx ts-node --transpile-only ./wiremock/resetStubs.ts",
     "start-test-wiremock": "docker-compose -f docker-compose-test.yml up -d",
     "generate-types": "script/generate-types",
+    "form-page:generate": "npx ts-node --transpile-only ./server/form-pages/utils/generator.ts",
```
Comment (pezholio) [inline package.json:36]: The generator has somewhat gone out of whack with the way form pages work now, so you may want to get rid of this commit

File: wiremock/applicationStubs.ts [hunk 3]
```diff
@@ -1,6 +1,5 @@
 import { guidRegex } from './index'
 
-import applicationSummaryFactory from '../server/testutils/factories/applicationSummary'
```
Comment (pezholio) [inline wiremock/applicationStubs.ts:3]: I'm not entirely sure how up to date these stubs are, as we no longer use them in dev, preferring to bring up the whole stack

**Review by pezholio — APPROVED** [review]
LGTM - just some minor things that you may want to bring in / delete

**Comment by bryangaledxw** [comment]
[How can we re-use the CAS1 Apply solution for CAS3's need](https://trello.com/c/OmMhGuVF/767-how-can-we-re-use-the-cas1-apply-solution-for-cas3s-need)

**Comment by bryangaledxw** [comment]
Thanks for the review, and cheers for the heads up on those. Will leave as is for this PR - I kinda want this as a line-in-the-sand that we're up to date with a specific known point (https://github.com/ministryofjustice/hmpps-approved-premises-ui/pull/536)

**Pull request merged**